File: js/data_structures/db_link_list.js
const LinkList = require('./link_list')

// 双向链表结构
class DBLinkList extends LinkList {
    constructor() {
        // 继承普通链表, 注入双向节点
        super({ LinkNode })
    }
    insert(newItem, item = 'head') {
        const curNode = this.find(item);
        const newNode = new this.LinkNode(newItem);
        /**
         * head -> preNode -> curNode -> nextNode
         * head -> preNode -> curNode -> newNode -> nextNode
         */
        if (curNode.next) {
            // 下一个的上一级索引也是需要修改的，不然任意插入会有问题
            curNode.next.prev = newNode
        }
        newNode.next = curNode.next;
        newNode.prev = curNode;
        curNode.next = newNode;
    }
    remove(item) {
        const curNode = this.find(item);
        if (curNode.next != null) {
            /**
             * preNode -> curNode -> nextNode -> nextNextNode
             * preNode -> nextNode -> nextNextNode
             */
            curNode.prev.next = curNode.next
            curNode.next.prev = curNode.prev
            // gc ?
            curNode.next = null
            curNode.prev = null
        } else {
            curNode.prev.next = null
            curNode.prev = null
            // curNode = null
        }
    }
    // 查找最后一个元素
    findLastOne() {
        let curNode = this.head;
        while(curNode.next) {
            curNode = curNode.next
        }
        return curNode
    }
    // 倒序遍历
    toReverseArr() {
        let curNode = this.findLastOne();
        const reverseArr = [];
        while(curNode.prev) {
            reverseArr.push(curNode.item)
            curNode = curNode.prev;
        }
        return reverseArr
    }
}

// 双向链表节点
class LinkNode {
    constructor(item = 'head') {
        this.item = item;
        // 存在prev 节点和 next节点
        this.prev = null;
        this.next = null;
    }
}

module.exports = DBLinkList

File: py/data_structures/db_linklist.py
# 双向链表
# 即使是在同级目录下，也应该使用相对目录的import
from data_structures.linklist import LinkList


class DBLinkList(LinkList):
    def __init__(self):
        super().__init__(node=Node)

    # 修改插入方法
    # head -> cur_node -> next_node
    # head -> cur_node -> new_node -> next_node
    def insert(self, new_item, item='head'):
        cur_node = self.find(item)
        new_node = self.ListNode(new_item)
        new_node.next = cur_node.next
        new_node.prev = cur_node
        cur_node.next = new_node

    # 删除节点
    # head -> cur_node -> next_node
    # head -> next_node
    def remove(self, item):
        target_node = self.find(item)
        target_node.prev.next = target_node.next
        target_node.next.prev = target_node.prev
        target_node.next = None
        target_node.prev = None

    # 查找最后一个节点
    def find_last_one(self):
        cur_node = self.head
        while cur_node.next is not None:
            cur_node = cur_node.next
        return cur_node

    # 倒序输出
    def to_reverse_arr(self):
        cur_node = self.find_last_one()
        arr = []
        while cur_node.prev is not None:
            arr.append(cur_node.item)
            cur_node = cur_node.prev
        return arr

# 双向节点
class Node(object):
    def __init__(self, item='head'):
        self.item = item
        self.prev = None
        self.next = None
